include/extractor/guidance/mergable_road_detector.hpp

```diff
@@ -0,0 +1,149 @@
+    bool NameIDsAreCompatible(const NameID lhs, const NameID rhs) const;
+
+    /* When it comes to merging roads, we need to find out if two ways actually represent the
+     * same road. This check tries to identify roads which are the same road in opposite directions
+     */
+    bool RoadDataIsCompatible(const util::NodeBasedEdgeData &lhs_edge_data,
+                              const util::NodeBasedEdgeData &rhs_edge_data) const;
+
+    /* Detect traffic loops.
+     * Since OSRM cannot handle loop edges, we cannot directly see a connection between a node and
+     * itself. We need to skip at least a single node in between. */
+    bool IsTrafficLoop(const NodeID intersection_node, const MergableRoadData &road) const;
+
+    /* Detetor to check if we are looking at roads splitting up just prior to entering an
+     * intersection:
+     *
+     *        c
+     *      / |
+     * a -<   |
+     *      \ |
+     *        b
+     *
+     * A common scheme in OSRM is that roads spit up in separate ways when approaching an
+     * intersection. This detector tries to detect these narrow triangles which usually just offer a
+     * small island for pedestrians in the middle. */
+    bool IsNarrowTriangle(const NodeID intersection_node,
+                          const MergableRoadData &lhs,
+                          const MergableRoadData &rhs) const;
+
+    /* Detector to check for whether two roads are following the same direction.
+     * If roads don't end up right at a connected intersection, we could look at a situation like
+     *
+     *      __________________________
+     *     /
+     * ---
+     *     \__________________________
+     *
+     * This detector tries to find out about whether two roads are parallel after the separation */
+    bool HaveSameDirection(const NodeID intersection_node,
+                           const MergableRoadData &lhs,
+                           const MergableRoadData &rhs) const;
+
+    /* Detector for small traffic islands. If a road is splitting up, just to connect again later,
+     * we don't wan't to have this information within our list of intersections/possible turn
+     * locations.
+     *
+     *     ___________
+     * ---<___________>-----
+     *
+     *
+     * Would feel just like a single straight road to a driver and should be represented as such in
+     * our engine.*/
+    bool ConnectAgain(const NodeID intersection_node,
+                      const MergableRoadData &lhs,
+                      const MergableRoadData &rhs) const;
+
+    /* A negative detector, preventing a merge, trying to detect link roads between two main roads.
+     *
+     *  d - - - - - - - - e - f
+     *             . / '
+     *  a - - - b - - - - - - c
+     *
+     * The detector wants to prevent merges that are connected to `b-e` */
+    bool IsLinkRoad(const NodeID intersection_node, const MergableRoadData &road) const;
```

Comment:
Well All of these are kind of specifically tailored to what we can merge into a single road. If these detectors trigger, all other handlers will not even see the respective roads (since they are merged into other ones). It might be that we can re-use them, for right now I would let them remain here. The worst thing that might happen is that we have to make this handler selection here accessible by making all private functions public.